source/dofs/dof_handler.cc

```diff
@@ -2569,104 +2551,72 @@ DoFHandler<dim, spacedim>::distribute_dofs(
-        .load_user_flags(user_flags);
     }
+
+  // Adjust size of levels to the triangulation. Note that we still have to
+  // allocate space for all degrees of freedom on this mesh (including ghost
+  // and cells that are entirely stored on different processors), though we
+  // may not assign numbers to some of them (i.e. they will remain at
+  // invalid_dof_index). We need to allocate the space because we will want
+  // to be able to query the dof_indices on each cell, and simply be told
+  // that we don't know them on some cell (i.e. get back invalid_dof_index)
+  if (hp_capability_enabled)
+    internal::hp::DoFHandlerImplementation::Implementation::reserve_space(
+      *this);
   else
-    {
-      // delete all levels and set them up newly. note that we still have to
-      // allocate space for all degrees of freedom on this mesh (including ghost
-      // and cells that are entirely stored on different processors), though we
-      // may not assign numbers to some of them (i.e. they will remain at
-      // invalid_dof_index). We need to allocate the space because we will want
-      // to be able to query the dof_indices on each cell, and simply be told
-      // that we don't know them on some cell (i.e. get back invalid_dof_index)
-      this->clear_space();
-      object_dof_indices.resize(this->tria->n_levels());
-      object_dof_ptr.resize(this->tria->n_levels());
-      cell_dof_cache_indices.resize(this->tria->n_levels());
-      cell_dof_cache_ptr.resize(this->tria->n_levels());
-      internal::DoFHandlerImplementation::Implementation::reserve_space(*this);
-
-      // hand things off to the policy
-      this->number_cache = this->policy->distribute_dofs();
-
-      // initialize the block info object only if this is a sequential
-      // triangulation. it doesn't work correctly yet if it is parallel
-      if (dynamic_cast<
-            const parallel::DistributedTriangulationBase<dim, spacedim> *>(
-            &*this->tria) == nullptr)
-        this->block_info_object.initialize(*this, false, true);
-    }
+    internal::DoFHandlerImplementation::Implementation::reserve_space(*this);
+
+  // undo the subdomain modification
+  if (shared_tria != nullptr && shared_tria->with_artificial_cells())
+    for (const auto &cell : shared_tria->active_cell_iterators())
+      cell->set_subdomain_id(saved_subdomain_ids[cell->active_cell_index()]);
+
+  // hand the actual work over to the policy
+  this->number_cache = this->policy->distribute_dofs();
+
+  // do some housekeeping: compress indices
+  // if(hp_capability_enabled)
+  //   {
+  //     Threads::TaskGroup<> tg;
+  //     for (int level = this->levels_hp.size() - 1; level >= 0; --level)
+  //       tg += Threads::new_task(
+  //         &dealii::internal::hp::DoFLevel::compress_data<dim, spacedim>,
+  //         *this->levels_hp[level],
+  //         this->fe_collection);
+  //     tg.join_all();
+  //   }
```

Comment:
This reminds me that might want to compress the indices again. As I see, it is on the TODO list of https://github.com/dealii/dealii/issues/10333#issue-623676096.